### PR TITLE
[MC-1262] InsightIDR plugin new actions

### PR DESCRIPTION
**Pull request by dsliwinski-r7**
### Description

Describe the proposed changes:

  - Add new actions `List Comments`, `Create Comment`, `Delete Comment`, `List Attachments`, `Upload Attachment`, `Download Attachment`, `Delete Attachment`, `Get Attachment Information`

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: create_comment\nCreating a comment for rrn:investigation:us:3af18a2c-432f-4bce-af67-93711c509704:investigation:59VSY1IDWA7W...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "comment": {
        "created_time": "2022-09-23T10:02:26.417Z",
        "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:comment:4PM8Z732W7BL",
        "target": "rrn:investigation:us:3af18a2c-432f-4bce-af67-93711c509704:investigation:59VSY1IDWA7W",
        "creator": {
          "type": "USER",
          "name": "Example User"
        },
        "body": "test",
        "visibility": "PUBLIC",
        "attachments": [
          {
            "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:UQHQX90POXQE",
            "creator": {
              "type": "USER",
              "name": "Example User"
            },
            "created_time": "2022-09-20T08:55:53.353Z",
            "file_name": "test.json",
            "mime_type": "text/plain",
            "size": 4,
            "scan_status": "CLEAN"
          }
        ]
      },
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/create_comment.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: delete_attachment\nDeleting the rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:SVJOR13GOKMG attachment...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/delete_attachment.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: delete_comment\nDeleting the rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:comment:4PM8Z732W7BL comment...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/delete_comment.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: download_attachment\nDownloading the rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:SVJOR13GOKMG attachment...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "attachment_content": "dGVzdA==",
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/download_attachment.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: get_attachment_information\nGetting the attachment information for rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:MU56PS5E8RO3...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "attachment": {
        "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:MU56PS5E8RO3",
        "creator": {
          "type": "USER",
          "name": "Example User"
        },
        "created_time": "2022-09-20T09:39:12.766Z",
        "file_name": "test.txt",
        "mime_type": "text/plain",
        "size": 4,
        "scan_status": "CLEAN"
      },
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/get_attachment_information.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: list_attachments\nListing the attachments for rrn:investigation:us:3af18a2c-432f-4bce-af67-93711c509704:investigation:59VSY1IDWA7W...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "attachments": [
        {
          "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:F264R9KSHMEI",
          "creator": {
            "type": "USER",
            "name": "Example User"
          },
          "created_time": "2022-08-19T13:00:58.645Z",
          "file_name": "image.PNG",
          "mime_type": "image/png",
          "size": 23298,
          "scan_status": "CLEAN"
        },
        {
          "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:UQHQX90POXQE",
          "creator": {
            "type": "USER",
            "name": "Example User"
          },
          "created_time": "2022-09-20T08:55:53.353Z",
          "file_name": "test.json",
          "mime_type": "text/plain",
          "size": 4,
          "scan_status": "CLEAN"
        }
      ],
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/list_attachments.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: list_comments\nListing the comments for rrn:investigation:us:3af18a2c-432f-4bce-af67-93711c509704:investigation:59VSY1IDWA7W...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "comments": [
        {
          "created_time": "2022-09-22T09:10:19.025Z",
          "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:comment:Q6K5VOQMAKL7",
          "target": "rrn:investigation:us:3af18a2c-432f-4bce-af67-93711c509704:investigation:59VSY1IDWA7W",
          "creator": {
            "type": "USER",
            "name": "Example User"
          },
          "visibility": "PUBLIC",
          "attachments": [
            {
              "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:UQHQX90POXQE",
              "creator": {
                "type": "USER",
                "name": "Example User"
              },
              "created_time": "2022-09-20T08:55:53.353Z",
              "file_name": "test.json",
              "mime_type": "text/plain",
              "size": 4,
              "scan_status": "CLEAN"
            }
          ]
        }
      ],
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/list_comments.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Rapid7 InsightIDR:4.1.0. Step name: upload_attachment\nUploading an attachment...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "attachment": [
        {
          "rrn": "rrn:collaboration:us:3af18a2c-432f-4bce-af67-93711c509704:attachment:SVJOR13GOKMG",
          "creator": {
            "type": "USER",
            "name": "Example User"
          },
          "created_time": "2022-09-23T10:08:08.071Z",
          "file_name": "test.txt",
          "mime_type": "text/plain",
          "size": 4,
          "scan_status": "PENDING"
        }
      ],
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
} 
```

<summary>
docker run --rm -i rapid7/rapid7_insightidr:4.1.0 run < tests/upload_attachment.json
</summary>
</details>

### CODE REVIEW

**Review by mgrandys-r7 — APPROVED** [review]

**Review by jkaluza-rapid7 — APPROVED** [review]

File: plugins/rapid7_insightidr/unit_test/test_create_comment.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    def test_create_comment(self, mock_request, name, target, body, attachments, expected):
+        actual = self.action.run({Input.TARGET: target, Input.BODY: body, Input.ATTACHMENTS: attachments})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("create_comment_bad").get("parameters"))
+    def test_create_comment_bad(self, mock_request, name, target, body, attachments, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.TARGET: target, Input.BODY: body, Input.ATTACHMENTS: attachments})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_create_comment.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.TARGET: target, Input.BODY: body, Input.ATTACHMENTS: attachments})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

File: plugins/rapid7_insightidr/unit_test/test_delete_attachment.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    def test_delete_attachment(self, mock_request, name, attachment_rrn, expected):
+        actual = self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("delete_comment_bad").get("parameters"))
+    def test_delete_attachment_bad(self, mock_request, name, attachment_rrn, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_delete_attachment.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

File: plugins/rapid7_insightidr/unit_test/test_delete_comment.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    def test_delete_comment(self, mock_request, name, comment_rrn, expected):
+        actual = self.action.run({Input.COMMENT_RRN: comment_rrn})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("delete_comment_bad").get("parameters"))
+    def test_delete_comment_bad(self, mock_request, name, comment_rrn, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.COMMENT_RRN: comment_rrn})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_delete_comment.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.COMMENT_RRN: comment_rrn})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

File: plugins/rapid7_insightidr/unit_test/test_download_attachment.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    def test_download_attachment(self, mock_request, name, attachment_rrn, expected):
+        actual = self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("download_attachment_bad").get("parameters"))
+    def test_download_attachment_bad(self, mock_request, name, attachment_rrn, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_download_attachment.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

File: plugins/rapid7_insightidr/unit_test/test_get_attachment_information.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    def test_get_attachment_information(self, mock_request, name, attachment_rrn, expected):
+        actual = self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("get_attachment_information_bad").get("parameters"))
+    def test_get_attachment_information_bad(self, mock_request, name, attachment_rrn, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_get_attachment_information.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.ATTACHMENT_RRN: attachment_rrn})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

File: plugins/rapid7_insightidr/unit_test/test_list_attachments.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    def test_list_attachments(self, mock_request, name, target, index, size, expected):
+        actual = self.action.run({Input.TARGET: target, Input.INDEX: index, Input.SIZE: size})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("list_attachments_bad").get("parameters"))
+    def test_list_attachments_bad(self, mock_request, name, target, index, size, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.TARGET: target, Input.INDEX: index, Input.SIZE: size})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_list_attachments.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.TARGET: target, Input.INDEX: index, Input.SIZE: size})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

File: plugins/rapid7_insightidr/unit_test/test_list_comments.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    def test_list_comments(self, mock_request, name, target, index, size, expected):
+        actual = self.action.run({Input.TARGET: target, Input.INDEX: index, Input.SIZE: size})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("list_comments_bad").get("parameters"))
+    def test_list_comments_bad(self, mock_request, name, target, index, size, cause, assistance):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.TARGET: target, Input.INDEX: index, Input.SIZE: size})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/unit_test/test_list_comments.py:31]:
```suggestion
        with self.assertRaises(PluginException) as error:
            self.action.run({Input.TARGET: target, Input.INDEX: index, Input.SIZE: size})
        self.assertEqual(error.exception.cause, cause)
        self.assertEqual(error.exception.assistance, assistance)
```

**Review by igorski-r7 — CHANGES_REQUESTED** [review]
Small changes 😃

**Comment by dsliwinski-r7** [comment]
> Small changes 😃

@igorski-r7 Done 😃

**Review by igorski-r7 — APPROVED** [review]

**Pull request merged**